[#37] 전체 사용자 프로필 목록 불러오기 기능 구현

msugo1 opened the pull request:
## 📖 Model 관련

### 📑 UserProfiles

- 해당 요청을 처리한 결과를 담을 객체를 생성하는데 사용할 클래스

## 📖 Controller 계층 관련

### 📑 ProfileController

- 전체 사용자 프로필 조회에 대한 요청을 매핑할 `getUserProfiles` 메소드 추가

## 📖 Service 계층 관련

### 📑 ProfileService

- 프로필 조회에 대한 요청을 Mapper 계층으로 전달해 줄 비즈니스 로직을 담은 `getUserProfiles` 메소드 추가

- 메인 페이지 캐싱 추가를 위해 `@Cacheable` 어노테이션과 커서 값이 주어져 있지 않을 때 캐싱하도록 `conditions = '#pagination.cursor == null'`조건 값 추가

### 📑 UserService

- 현재 메인 페이지는 유저의 가입 순서(id, surrogate key)를 기준으로 내림차순 정렬

- 따라서 새로운 유저가 가입했을 때 업데이트 내역이 반영될 수 있도록 유저 프로필들 캐싱내용을 제거할 @CacheEvict 추가

## 📖 Mapper 계층 관련

### 📑 ProfileMapper

- 비즈니스 단에서 전달 받은 내용을 바탕으로 DB에서 데이터를 불러올 `getUserProfilesList` 메소드 추가

### 📑 ProfileMapper.xml

- 위의 로직을 처리할 쿼리문 추가

- `resultMap`과 `assosiation`을 사용해서 기본 쿼리문과 각 유저가 받은 추천 글의 개수를 불러올 서브 쿼리문을 한번에 처리하도록 하여 DB에 연결 횟수 줄이도록 구성

## 📖 Configuration 관련

### 📑 RedisCacheConfig

- 메인 페이지를 대상으로 하는 캐싱 설정 값 추가

- 해당 캐시의 지속 시간은 5분으로 산정 (추후 변경 가능)

### 📑 Util 관련

- 메인 페이지의 캐싱에 활용할 키/밸류 값에 대한 상수(`MAIN_PAGE_KEY`, `MAIN_PAGE_VALUE`) 추가

Commit:
[#37] 전체 사용자 프로필 조회기능 추가

Model 관련

* UserProfileList

- 해당 요청을 처리한 결과를 담을 객체를 생성하는데 사용할 클래스

Controller 계층 관련

* ProfileController

- 전체 사용자 프로필 조회에 대한 요청을 매핑할 `getUserProfileList`
  메소드 추가

Service 계층 관련

* ProfileService

- 프로필 조회에 대한 요청을 Mapper 계층으로 전달해 줄 비즈니스 로직을
  담은 `getUserProfilesList` 메소드 추가

Mapper 계층 관련

* ProfileMapper

- 비즈니스 단에서 전달 받은 내용을 바탕으로 DB에서 데이터를 불러올 `getUserProfilesList` 메소드 추가

* ProfileMapper.xml

- 위의 로직을 처리할 쿼리문 추가

- `resultMap`과 `assosiation`을 사용해서 기본 쿼리문과 각 유저가 받은 추천 글의 개수를
  불러올 서브 쿼리문을 처리하도록 구성

Commit:
Merge branch 'feature/36' into feature/37

Commit:
[#37] 사용자 프로필 불러오기 메인 페이지 캐싱 추가

Configuration 관련

* RedisCacheConfig

- 메인 페이지를 대상으로 하는 캐싱 설정 값 추가

- 해당 캐시의 지속 시간은 5분으로 산정 (추후 변경 가능)

Controller 계층 관련

* ProfileController

- 전체 사용자 목록을 불러오는 메소드의 이름을 `getUserProfilesList -> getUserProfiles`로 변경

Service 계층 관련

* ProfileService

- 메소드 이름을 `getUserProfilesList -> getUserProfiles`로 변경

- 메인 페이지 캐싱 추가를 위해 `@Cacheable` 어노테이션과 커서 값이
  주어져 있지 않을 때 캐싱하도록 `conditions = '#pagination.cursor ==
  null'`설정 값 추가

* UserService

- 현재 메인 페이지는 유저의 가입 순서(id, surrogate key)를 기준으로 내림차순 정렬

- 따라서 새로운 유저가 가입했을 때 업데이트 내역이 반영될 수 있도록
  유저 프로필들 캐싱내용을 제거할 @CacheEvict 추가

Mapper 계층 관련

* ProfileMapper/-.xml

- 메소드 및 클래스 이름 변경 내역 반영

Model 관련

* UserProfiles

- 클래스 이름 변경 `UserProfileList -> UserProfiles`

Util 관련

- 메인 페이지의 캐싱에 활용할 키/밸류 값에 대한 상수(`MAIN_PAGE_KEY`, `MAIN_PAGE_VALUE`) 추가

Commit:
[#37] feature/36에 대한 변경사항을 feature/37에 반영

f-lab-dev commented on src/main/java/me/soo/helloworld/service/ProfileService.java:
```diff
@@ -41,6 +42,12 @@ public UserProfile getUserProfile(String userId) {
     }
 
+    @Transactional(readOnly = true)
+    @Cacheable(key = MAIN_PAGE_KEY, value = MAIN_PAGE_VALUE, condition = "#pagination.cursor == null", cacheManager = REDIS_CACHE_MANAGER)
```
첫 페이지만 캐싱하시려는 것 같은데 다른 페이지는 캐싱하지 않는 이유가 있으실까요?

msugo1 replied:
- 처음에는 사용자의 프로필을 나열해서 보여주는 첫 화면을 사용자 추천 알고리즘을 적용하여 무작위 결과가 생성되게 한 후 이 부분을 캐싱해서 캐시 유효시간 동안 사용자에게 같은 결과를 보여주는 것을 구상했었습니다. 

- 그리고 첫 페이지 이후에는 스크롤을 내리면서 아마 cursor 값을 key로 해서 캐싱을 하게 될텐데 사용자의 추가/삭제 시, 커서가 가리키는 포인트의 시작 점이 변동 되어 다른 key 값을 가지게 되므로 캐싱 처리를 하면 똑같은 내용이 여러 번 반복되서 캐싱될 문제가 있다고 생각했습니다.

- 지금은 방향을 좀 바꿔서 메인 화면에 최근에 가입한 사용자의 프로필부터 보여지도록 할 예정인데요. cursor를 기준으로 키를 설정하되, 지속시간을 짧게하고(15초 정도) 레디스 저장소 내부에서 Eviction을 적절히 설정해주면 다른 키 값 안에 중복되는 밸류가 들어가더라도 사용자의 추가/삭제가 이루어지지 않는 시점에는 사용자들이 cursor 값을 가지고 조회할 때 응답속도를 높일 수 있을 것 같은데 괜찮을까요?

f-lab-dev replied:
커서를 기준으로 캐싱을 설정하면 커서 이전의 데이터들을 보여주는것이기 때문에 캐싱시에 정합성 문제가 발생 안하지 않을까요?

f-lab-dev replied:
네 그러면 괜찮을 것 같네요~

Commit:
[#37] feature/36의 변경사항을 feature/37에 반영

Commit:
[#37] 타 사용자의 차단 내역이 조회에 반영되게 수정

* ProfileController

- 현재 사용자의 아이디를 파라미터로 넘겨서 조회하려는 사용자/사용자들이
  차단 목록에 현재 사용자를 가지고 있는지 판별하는데 활용

* ProfileService

- Mapper 계층에 프로필 정보 조회 시 현재 사용자의 아이디도 함께
  전달하도록 변경

* ProfileMapper

- DB에 프로필 정보 조회 시 현재 사용자의 아이디 활용

* ProfileMapper.xml

- 현재 사용자가 조회하려는 대상의 차단 목록에 포함되어 있는지 조회하는
  쿼리를 WHERE 조건에 추가하여 차단된 사용자들이 조회되지 않도록 쿼리
  변경

* blockUserMapper.xml

- 인덱스를 제대로 활용하기 위해서는 인덱스가 등록된 칼럼의 순서대로
  조회되어야 한다는 특징이 있음

- 기존 칼럼 (userId, blockId)의 경우는 다른 사용자의 차단 목록을 조회하는 쿼리를 처리할 때 칼럼 순서가 맞지 않아서 인덱스를 제대로 활용하지 못한다는 문제점이 존재했음

- 인덱스를 활용할 수 있게 칼럼의 순서를 변경 및 해당 내용을 쿼리에 반영(blockId, blockedBy)

Commit:
[#37] 프로필 조회 캐싱 옵션 추가

- 다수의 사용자의 프로필을 조회하는 경우 메인 페이지(cursor가 지정되어
  있지 않은) 이외에도 cursor를 키로 활용하여 캐싱하도록 설정 추가

- cursor 의 경우 회원이 새로 가입하거나 탈퇴하는 경우 변동사항이
  발생하여 같은 내용이 다수 캐싱되는 문제가 있을 수 있으므로 지속시간을
  짧게 설정

Commit:
[#37] LocalDate Deserialisation 예외 수정

- Redis에서 캐싱한 데이터를 불러올 때 LocalDate가 제대로 역직렬화되지 않던 문제 발생

- `@JsonSerialize/Deserialize` 어노테이션을 활용해 직렬화/역직렬화
  클래스 설정

Commit:
[#37] 피드백 반영

* feature/36의 변동사항 반영

- Merge branch 'feature/36' into feature/37

* 전체 프로필 캐싱 설정 변경

- 기존: 첫 페이지만 캐싱

- 변경: 첫 페이지는 먼저 캐싱해 놓고 이후 커서를 기준으로 15초 간격으로
  캐싱하면서 응답속도를 높이도록 변경. 레디스 저장소에는 LFU(Least Frequently Used) 캐시 Eviction 설정을 통해 자주 사용되지 않는 캐시를 제거함으로써 효율성을 높이고자 함

f-lab-dev commented on src/main/resources/mappers/ProfileMapper.xml:
```diff
@@ -73,18 +75,19 @@
-        SELECT id, userId, aboutMe,
-               profileImageName, profileImagePath
+    <select id="getUserProfiles" parameterType="map"
+            resultMap="getUserProfilesMap">        SELECT id, userId, aboutMe,
```
여기 쿼리쪽에 개행 한번 해주시면 좋을 것 같아요~

f-lab-dev reviewed (approved):
오타부분만 해결하시고 머지하시면 될 것 같아요~

Commit:
[#37] 피드백 반영

- ProfileMapper 쿼리 오타 수정
